### PR TITLE
Bring back logging of discarded samples

### PR DESCRIPTION
**Pull request by beorn7**
But only on DEBUG level.

Also, count and report the two cases of out-of-order timestamps on the
one hand and same timestamp but different value on the other hand
separately.

@juliusv @richih @jjneely

### CODE REVIEW

**Comment by RichiH** [comment]
Thanks, will try tomorrow.

Just to make sure: you are counting without DEBUG, correct?

File: storage/local/storage.go [hunk 1]
```diff
@@ -242,12 +242,15 @@ func NewMemorySeriesStorage(o *MemorySeriesStorageOptions) Storage {
...
-			Help:      "The total number of samples that were discarded because their timestamps were at or before the last received sample for a series.",
-		}),
+		discardedSamplesCount: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Subsystem: subsystem,
+				Name:      "out_of_order_samples_total",
+				Help:      "The total number of samples that were discarded because their timestamps were at or before the last received sample for a series.",
+			},
+			[]string{discardReasonLabel},
```
Comment (brian-brazil) [inline storage/local/storage.go]: You should initialize both time series to 0.

**Comment by beorn7** [comment]
Yes, the metrics are always there. So if you see them jump up, you can enable debug logging and get log-spammed.

File: storage/local/storage.go (hunk 1 above)
Comment (beorn7) [inline storage/local/storage.go]: Oh yeah, I did that, and also initialized the other metric vecs in memorySeriesStorage for consistency.

File: storage/local/storage.go [hunk 2]
```diff
@@ -276,6 +279,25 @@ func NewMemorySeriesStorage(o *MemorySeriesStorageOptions) Storage {
 			Help:      "1 if the storage is in rushed mode, 0 otherwise. In rushed mode, the system behaves as if the persistence_urgency_score is 1.",
 		}),
 	}
+
+	// Initialize metric vectors.
+	// TODO(beorn7): Rework once we have a utility function for it in golang_client.
```
Comment (juliusv) [inline storage/local/storage.go]: "golang_client" -> "client_golang"

File: storage/local/storage.go [hunk 3]
```diff
@@ -577,9 +599,16 @@ func (s *memorySeriesStorage) DropMetricsForFingerprints(fps ...model.Fingerprin
...
-// ErrOutOfOrderSample is returned if a sample has a timestamp before the latest
-// timestamp in the series it is appended to.
-var ErrOutOfOrderSample = fmt.Errorf("sample timestamp out of order")
+var (
+	// ErrOutOfOrderSample is returned if a sample has a timestamp before the latest
+	// timestamp in the series it is appended to.
+	ErrOutOfOrderSample = fmt.Errorf("sample timestamp out of order")
+	// ErrDuplicateSampleForTimestamp is returned if a sample has the same
+	// timestamp as the latest sample in the series it is appended to but a
+	// different value. (Appending an identical sample is a no-op does not
```
Comment (juliusv) [inline storage/local/storage.go]: "does" -> "and does"

**Comment by juliusv** [comment]
:+1: otherwise

**Comment by fabxc** [comment]
👍

**Comment by beorn7** [comment]
All addressed. Merging on green...

**Pull request merged**